### PR TITLE
[SPARK-13180] Protect against SessionState being null when accessing HiveClientImpl#conf

### PR DESCRIPTION
**Pull request by tedyu**
See http://search-hadoop.com/m/q3RTtFoTDi2HVCrM1 for related stack trace

```
java.lang.NullPointerException
at org.apache.spark.sql.hive.client.ClientWrapper.conf(ClientWrapper.scala:205)
at org.apache.spark.sql.hive.HiveContext.hiveconf$lzycompute(HiveContext.scala:552)
at org.apache.spark.sql.hive.HiveContext.hiveconf(HiveContext.scala:551)
at org.apache.spark.sql.hive.HiveContext$$anonfun$configure$1.apply(HiveContext.scala:538)
at org.apache.spark.sql.hive.HiveContext$$anonfun$configure$1.apply(HiveContext.scala:537)
at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
at scala.collection.immutable.List.foreach(List.scala:318)
at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
at scala.collection.AbstractTraversable.map(Traversable.scala:105)
at org.apache.spark.sql.hive.HiveContext.configure(HiveContext.scala:537)
at org.apache.spark.sql.hive.HiveContext.metadataHive$lzycompute(HiveContext.scala:250)
at org.apache.spark.sql.hive.HiveContext.metadataHive(HiveContext.scala:237)
at org.apache.spark.sql.hive.HiveContext$$anon$2.<init>(HiveContext.scala:457)
at org.apache.spark.sql.hive.HiveContext.catalog$lzycompute(HiveContext.scala:457)
at org.apache.spark.sql.hive.HiveContext.catalog(HiveContext.scala:456)
at org.apache.spark.sql.hive.HiveContext$$anon$3.<init>(HiveContext.scala:473)
at org.apache.spark.sql.hive.HiveContext.analyzer$lzycompute(HiveContext.scala:473)
at org.apache.spark.sql.hive.HiveContext.analyzer(HiveContext.scala:472)
at org.apache.spark.sql.execution.QueryExecution.assertAnalyzed(QueryExecution.scala:34)
at org.apache.spark.sql.DataFrame.<init>(DataFrame.scala:133)
at org.apache.spark.sql.DataFrame$.apply(DataFrame.scala:52)
at org.apache.spark.sql.SQLContext.baseRelationToDataFrame(SQLContext.scala:442)
at org.apache.spark.sql.DataFrameReader.jdbc(DataFrameReader.scala:223)
at org.apache.spark.sql.DataFrameReader.jdbc(DataFrameReader.scala:146)
```

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 1]
```diff
@@ -134,7 +134,18 @@ private[hive] class HiveClientImpl(
   }
 
   /** Returns the configuration for the current session. */
-  def conf: HiveConf = SessionState.get().getConf
+  def conf: HiveConf = {
```
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: This is an exact copy of https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveQl.scala#L171-L182.

Can you refactor this?

**Comment by tedyu** [comment]
Do you have suggestion on which class to put the common code in ?

Thanks

**Comment by hvanhovell** [comment]
@tedyu lets create a ```HiveConfUtil``` object in the client package, put the method in and use that in both places.

**Comment by tedyu** [comment]
This error surfaced:
```
[info]   2016-02-04 10:15:21.319 - stderr> Exception in thread "main" java.lang.LinkageError: loader constraint violation: when resolving method "org.apache.spark.sql.hive.      client.HiveConfUtil$.conf()Lorg/apache/hadoop/hive/conf/HiveConf;" the class loader (instance of org/apache/spark/sql/hive/client/IsolatedClientLoader$$anon$1) of the current    class, org/apache/spark/sql/hive/client/HiveClientImpl, and the class loader (instance of sun/misc/Launcher$AppClassLoader) for resolved class, org/apache/spark/sql/hive/client/ HiveConfUtil$, have different Class objects for the type /hadoop/hive/conf/HiveConf; used in the signature
```

**Comment by tedyu** [comment]
I wonder if rev 1 should be used:

the utility class only saves less than 10 lines of code
intricacies were introduced w.r.t. class loading

**Comment by tedyu** [comment]
@hvanhovell 
What do you think ?

**Comment by tedyu** [comment]
@davies 
Can you take a look ?

**Comment by davies** [comment]
@tedyu This seems like a workaround. Have you find the root cause? Maybe there is a place that we didn't set the SessionState for current thread.

**Comment by tedyu** [comment]
From HiveContext.scala , line 552 (the last line below):
```
   * 1. create a new SessionState for each HiveContext
   * 2. when the Hive session is first initialized, params in HiveConf will get picked up by the
   *    SQLConf.  Additionally, any properties set by set() or a SET command inside sql() will be
   *    set in the SQLConf *as well as* in the HiveConf.
   */
  @transient
  protected[hive] lazy val hiveconf: HiveConf = {
    val c = executionHive.conf
```
If I read the doc correctly, the above call is supposed to create new SessionState which is what the PR is doing.

**Comment by davies** [comment]
The doc is out-updated, HiveConf is created in HiveClientImp.scla LINE 107 (when created HiveClientImp obj).

**Comment by tedyu** [comment]
The user was using 1.6.0 where there was no HiveClientImp.scala

**Comment by rxin** [comment]
Thanks for the pull request. I'm going through a list of pull requests to cut them down since the sheer number is breaking some of the tooling we have. Due to lack of activity on this pull request, I'm going to push a commit to close it. Feel free to reopen it or create a new one. We can also continue the discussion on the JIRA ticket.

Note: Looks like this code path has been rewritten heavily and the problem might not apply anymore.

**Pull request closed without merging**